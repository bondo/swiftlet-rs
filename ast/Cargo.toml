[package]
name = "ast"
version.workspace = true
edition.workspace = true
authors.workspace = true

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
ariadne = "0.3.0"
pest.workspace = true
thiserror.workspace = true
yansi = "0.5.1"
