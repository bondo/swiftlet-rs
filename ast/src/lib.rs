mod error;
mod nodes;

pub use error::{SwiftletError, SwiftletErrorPrinter};
pub use nodes::*;
