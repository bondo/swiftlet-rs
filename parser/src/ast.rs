use tree_sitter::TreeCursor;

use crate::{InternalParserError, UserParseError};

#[derive(Debug, PartialEq)]
pub struct Program {
    pub exprs: Vec<Expr>,
}

impl Program {
    pub(crate) fn parse(
        cursor: &mut TreeCursor,
        source: &[u8],
    ) -> Result<(Self, Vec<UserParseError>), InternalParserError> {
        let node = cursor.node();
        debug_assert_eq!(node.kind(), "source_file");

        if !cursor.goto_first_child() {
            return Ok((
                Self { exprs: vec![] },
                vec![not_implemented_error(
                    "Program::parse::goto_first_child",
                    node,
                )],
            ));
        }

        let mut exprs: Vec<Expr> = vec![];
        let mut errors: Vec<UserParseError> = vec![];

        loop {
            let node = cursor.node();
            match node.kind() {
                "property_declaration" | "call_expression" => {
                    let (expr, mut errs) = Expr::parse(cursor, source)?;
                    if let Some(expr) = expr {
                        exprs.push(expr);
                    }
                    errors.append(&mut errs);
                }
                "comment" => (),
                "ERROR" => errors.push(parse_error(node)),
                _ => errors.push(not_implemented_error("Program", node)),
            }
            if !cursor.goto_next_sibling() {
                break;
            }
        }

        cursor.goto_parent();

        Ok((Self { exprs }, errors))
    }
}

#[derive(Debug, PartialEq)]
pub enum Expr {
    CallExpression(CallExpression),
    PropertyDeclaration(PropertyDeclaration),
    SimpleIdentifier(SimpleIdentifier),
    IntegerLiteral(i32),
}

impl Expr {
    fn parse(
        cursor: &mut TreeCursor,
        source: &[u8],
    ) -> Result<(Option<Self>, Vec<UserParseError>), InternalParserError> {
        let node = cursor.node();
        match node.kind() {
            "call_expression" => {
                let (expr, errs) = CallExpression::parse(cursor, source)?;
                Ok((expr.map(Self::CallExpression), errs))
            }
            "property_declaration" => {
                let (expr, errs) = PropertyDeclaration::parse(cursor, source)?;
                Ok((expr.map(Self::PropertyDeclaration), errs))
            }
            "simple_identifier" => Ok((
                Some(Self::SimpleIdentifier(SimpleIdentifier::parse(
                    cursor, source,
                )?)),
                vec![],
            )),
            "integer_literal" => node.utf8_text(source)?.parse().ok().map_or(
                Ok((
                    None,
                    vec![not_implemented_error("Expr::integer_literal", node)],
                )),
                |value| Ok((Some(Self::IntegerLiteral(value)), vec![])),
            ),
            "ERROR" => Ok((None, vec![parse_error(node)])),
            _ => Ok((None, vec![not_implemented_error("Expr", node)])),
        }
    }
}

#[derive(Debug, PartialEq)]
pub struct CallExpression {
    pub callee: SimpleIdentifier,
    pub arguments: CallSuffix,
}

impl CallExpression {
    fn parse(
        cursor: &mut TreeCursor,
        source: &[u8],
    ) -> Result<(Option<Self>, Vec<UserParseError>), InternalParserError> {
        let node = cursor.node();
        debug_assert_eq!(node.kind(), "call_expression");

        // todo!(
        //     "CallExpression::parse: {:#?}",
        //     node.children(cursor)
        //         .map(|n| format!("{kind}: {sexp}", kind = n.kind(), sexp = n.to_sexp()))
        //         .collect::<Vec<_>>()
        // );

        if !cursor.goto_first_child() {
            return Ok((
                None,
                vec![not_implemented_error(
                    "CallExpression::parse::goto_first_child",
                    node,
                )],
            ));
        }

        let mut errors: Vec<UserParseError> = vec![];

        let node = cursor.node();
        let (callee, mut errs) = match node.kind() {
            "simple_identifier" => (Some(SimpleIdentifier::parse(cursor, source)?), vec![]),
            "ERROR" => (None, vec![parse_error(node)]),
            _ => (None, vec![not_implemented_error("CallExpression", node)]),
        };
        errors.append(&mut errs);

        if !cursor.goto_next_sibling() {
            errors.push(not_implemented_error(
                "CallExpression::parse::goto_next_sibling",
                node.parent().unwrap(),
            ));
            cursor.goto_parent();
            return Ok((None, errors));
        }

        let node = cursor.node();
        let (arguments, mut errs) = match node.kind() {
            "call_suffix" => CallSuffix::parse(cursor, source)?,
            "ERROR" => (None, vec![parse_error(node)]),
            _ => (None, vec![not_implemented_error("CallExpression", node)]),
        };
        errors.append(&mut errs);

        if cursor.goto_next_sibling() {
            let node = cursor.node();
            errors.push(not_implemented_error(
                "CallExpression::parse::goto_next_sibling",
                node,
            ));
            cursor.goto_parent();
            return Ok((None, errors));
        }

        cursor.goto_parent();

        Ok(match (callee, arguments) {
            (Some(callee), Some(arguments)) => (Some(Self { callee, arguments }), errors),
            _ => (None, errors),
        })
    }
}

#[derive(Debug, PartialEq)]
pub enum CallSuffix {
    ValueArguments(ValueArguments),
}

impl CallSuffix {
    fn parse(
        cursor: &mut TreeCursor,
        source: &[u8],
    ) -> Result<(Option<Self>, Vec<UserParseError>), InternalParserError> {
        let node = cursor.node();
        debug_assert_eq!(node.kind(), "call_suffix");

        if !cursor.goto_first_child() {
            return Ok((
                None,
                vec![not_implemented_error(
                    "CallSuffix::parse::goto_first_child",
                    node,
                )],
            ));
        }

        let node = cursor.node();
        let (call_suffix, mut errors) = match node.kind() {
            "value_arguments" => {
                let (val, errs) = ValueArguments::parse(cursor, source)?;
                (Some(Self::ValueArguments(val)), errs)
            }
            "ERROR" => (None, vec![parse_error(node)]),
            _ => (None, vec![not_implemented_error("CallSuffix", node)]),
        };

        if cursor.goto_next_sibling() {
            let node = cursor.node();
            errors.push(not_implemented_error(
                "CallSuffix::parse::goto_next_sibling",
                node,
            ));
        }

        cursor.goto_parent();

        Ok((call_suffix, errors))
    }
}

#[derive(Debug, PartialEq)]
pub struct ValueArguments(pub(super) Vec<ValueArgument>);

impl ValueArguments {
    fn parse(
        cursor: &mut TreeCursor,
        source: &[u8],
    ) -> Result<(Self, Vec<UserParseError>), InternalParserError> {
        let node = cursor.node();
        debug_assert_eq!(node.kind(), "value_arguments");

        if !cursor.goto_first_child() {
            return Ok((
                Self(vec![]),
                vec![not_implemented_error(
                    "Program::parse::goto_first_child",
                    node,
                )],
            ));
        }

        let mut args: Vec<ValueArgument> = vec![];
        let mut errors: Vec<UserParseError> = vec![];

        let node = cursor.node();
        match node.kind() {
            "(" => (),
            "ERROR" => errors.push(parse_error(node)),
            _ => errors.push(not_implemented_error("ValueArguments", node)),
        };
        if !cursor.goto_next_sibling() {
            errors.push(not_implemented_error(
                "ValueArguments::parse::goto_next_sibling",
                node.parent().unwrap(),
            ));
            cursor.goto_parent();
            return Ok((Self(args), errors));
        }

        loop {
            let node = cursor.node();
            match node.kind() {
                "value_argument" => {
                    let (arg, mut errs) = ValueArgument::parse(cursor, source)?;
                    if let Some(arg) = arg {
                        args.push(arg);
                    }
                    errors.append(&mut errs);
                }
                ")" => {
                    if cursor.goto_next_sibling() {
                        let node = cursor.node();
                        errors.push(not_implemented_error(
                            "ValueArguments::parse::goto_next_sibling",
                            node,
                        ));
                    }
                    break;
                }
                "ERROR" => errors.push(parse_error(node)),
                _ => errors.push(not_implemented_error("ValueArguments", node)),
            }
            if !cursor.goto_next_sibling() {
                errors.push(not_implemented_error(
                    "ValueArguments::parse::goto_next_sibling",
                    node,
                ));
                break;
            }
        }

        cursor.goto_parent();

        Ok((Self(args), errors))
    }
}

#[derive(Debug, PartialEq)]
pub enum ValueArgument {
    SimpleIdentifier(SimpleIdentifier),
}

impl ValueArgument {
    fn parse(
        cursor: &mut TreeCursor,
        source: &[u8],
    ) -> Result<(Option<Self>, Vec<UserParseError>), InternalParserError> {
        let node = cursor.node();
        debug_assert_eq!(node.kind(), "value_argument");

        if !cursor.goto_first_child() {
            return Ok((
                None,
                vec![not_implemented_error(
                    "ValueArgument::parse::goto_first_child",
                    node,
                )],
            ));
        }

        let node = cursor.node();
        let (value_argument, mut errors) = match node.kind() {
            "simple_identifier" => (
                Some(Self::SimpleIdentifier(SimpleIdentifier::parse(
                    cursor, source,
                )?)),
                vec![],
            ),
            "ERROR" => (None, vec![parse_error(node)]),
            _ => (None, vec![not_implemented_error("ValueArgument", node)]),
        };

        if cursor.goto_next_sibling() {
            let node = cursor.node();
            errors.push(not_implemented_error(
                "ValueArgument::parse::goto_next_sibling",
                node,
            ));
        }

        cursor.goto_parent();

        Ok((value_argument, errors))
    }
}

#[derive(Debug, PartialEq)]
pub enum Qualifier {
    Var,
    Let,
}

#[derive(Debug, PartialEq)]
pub struct SimpleIdentifier {
    pub name: String,
}

impl SimpleIdentifier {
    fn parse(cursor: &mut TreeCursor, source: &[u8]) -> Result<Self, InternalParserError> {
        let node = cursor.node();
        debug_assert_eq!(node.kind(), "simple_identifier");

        Ok(Self {
            name: node.utf8_text(source)?.to_string(),
        })
    }
}

#[derive(Debug, PartialEq)]
pub enum Pattern {
    SimpleIdentifier(SimpleIdentifier),
}

impl Pattern {
    fn parse(
        cursor: &mut TreeCursor,
        source: &[u8],
    ) -> Result<(Option<Self>, Vec<UserParseError>), InternalParserError> {
        let node = cursor.node();
        debug_assert_eq!(node.kind(), "pattern");

        if !cursor.goto_first_child() {
            return Ok((
                None,
                vec![not_implemented_error(
                    "Pattern::parse::goto_first_child",
                    node,
                )],
            ));
        }

        let node = cursor.node();
        let (pattern, mut errors) = match node.kind() {
            "simple_identifier" => (
                Some(Self::SimpleIdentifier(SimpleIdentifier::parse(
                    cursor, source,
                )?)),
                vec![],
            ),
            "ERROR" => (None, vec![parse_error(node)]),
            _ => (None, vec![not_implemented_error("Pattern", node)]),
        };

        if cursor.goto_next_sibling() {
            let node = cursor.node();
            errors.push(not_implemented_error(
                "Pattern::parse::goto_next_sibling",
                node,
            ));
        }

        cursor.goto_parent();

        Ok((pattern, errors))
    }
}

#[derive(Debug, PartialEq)]
pub struct UserType {
    pub name: String,
}

impl UserType {
    fn parse(
        cursor: &mut TreeCursor,
        source: &[u8],
    ) -> Result<(Option<Self>, Vec<UserParseError>), InternalParserError> {
        let node = cursor.node();
        debug_assert_eq!(node.kind(), "user_type");

        if !cursor.goto_first_child() {
            return Ok((
                None,
                vec![not_implemented_error(
                    "UserType::parse::goto_first_child",
                    node,
                )],
            ));
        }

        let node = cursor.node();
        let (user_type, mut errors) = match node.kind() {
            "type_identifier" => (
                Some(Self {
                    name: node.utf8_text(source)?.to_string(),
                }),
                vec![],
            ),
            "ERROR" => (None, vec![parse_error(node)]),
            _ => (None, vec![not_implemented_error("UserType", node)]),
        };

        if cursor.goto_next_sibling() {
            let node = cursor.node();
            errors.push(not_implemented_error(
                "UserType::parse::goto_next_sibling",
                node,
            ));
        }

        cursor.goto_parent();

        Ok((user_type, errors))
    }
}

#[derive(Debug, PartialEq)]
pub enum TypeIdentifier {
    UserType(UserType),
}

impl TypeIdentifier {
    fn parse(
        cursor: &mut TreeCursor,
        source: &[u8],
    ) -> Result<(Option<Self>, Vec<UserParseError>), InternalParserError> {
        let node = cursor.node();
        debug_assert_eq!(node.kind(), "type_annotation");

        if !cursor.goto_first_child() {
            return Ok((
                None,
                vec![not_implemented_error(
                    "TypeIdentifier::parse::goto_first_child",
                    node,
                )],
            ));
        }

        let node = cursor.node();
        let mut errors = match node.kind() {
            ":" => vec![],
            "ERROR" => vec![parse_error(node)],
            _ => vec![not_implemented_error("TypeIdentifier", node)],
        };

        if !cursor.goto_next_sibling() {
            errors.push(not_implemented_error(
                "TypeIdentifier::parse::goto_next_sibling",
                node.parent().unwrap(),
            ));
            cursor.goto_parent();
            return Ok((None, errors));
        }

        let node = cursor.node();
        let (identifier, mut errs) = match node.kind() {
            "user_type" => {
                let (user_type, errs) = UserType::parse(cursor, source)?;
                (user_type.map(Self::UserType), errs)
            }
            "ERROR" => (None, vec![parse_error(node)]),
            _ => (None, vec![not_implemented_error("TypeIdentifier", node)]),
        };
        errors.append(&mut errs);

        if cursor.goto_next_sibling() {
            let node = cursor.node();
            errors.push(not_implemented_error(
                "TypeIdentifier::parse::goto_next_sibling",
                node,
            ));
        }

        cursor.goto_parent();

        Ok((identifier, errors))
    }
}

#[derive(Debug, PartialEq)]
pub struct PropertyDeclaration {
    pub qualifier: Qualifier,
    pub lhs: Pattern,
    pub r#type: Option<TypeIdentifier>,
    pub rhs: Box<Expr>,
}

impl PropertyDeclaration {
    fn parse(
        cursor: &mut TreeCursor,
        source: &[u8],
    ) -> Result<(Option<Self>, Vec<UserParseError>), InternalParserError> {
        let node = cursor.node();
        debug_assert_eq!(node.kind(), "property_declaration");

        if !cursor.goto_first_child() {
            return Ok((
                None,
                vec![not_implemented_error(
                    "PropertyDeclaration::parse::goto_first_child",
                    node,
                )],
            ));
        }

        let mut errors: Vec<UserParseError> = vec![];

        let node = cursor.node();
        let (qualifier, mut errs) = match node.kind() {
            "var" => (Some(Qualifier::Var), vec![]),
            "let" => (Some(Qualifier::Let), vec![]),
            "ERROR" => (None, vec![parse_error(node)]),
            _ => (
                None,
                vec![not_implemented_error("PropertyDeclaration", node)],
            ),
        };
        errors.append(&mut errs);

        if !cursor.goto_next_sibling() {
            errors.push(not_implemented_error(
                "PropertyDeclaration::parse::goto_next_sibling",
                node.parent().unwrap(),
            ));
            cursor.goto_parent();
            return Ok((None, errors));
        }

        let node = cursor.node();
        let (lhs, mut errs) = match node.kind() {
            "pattern" => Pattern::parse(cursor, source)?,
            "ERROR" => (None, vec![parse_error(node)]),
            _ => (
                None,
                vec![not_implemented_error("PropertyDeclaration", node)],
            ),
        };
        errors.append(&mut errs);

        if !cursor.goto_next_sibling() {
            errors.push(not_implemented_error(
                "PropertyDeclaration::parse::goto_next_sibling",
                node.parent().unwrap(),
            ));
            cursor.goto_parent();
            return Ok((None, errors));
        }

        let node = cursor.node();
        let (r#type, mut errs) = match node.kind() {
            "type_annotation" => TypeIdentifier::parse(cursor, source)?,
            "=" => (None, vec![]),
            "ERROR" => (None, vec![parse_error(node)]),
            _ => (
                None,
                vec![not_implemented_error("PropertyDeclaration", node)],
            ),
        };
        errors.append(&mut errs);

        if r#type.is_some() {
            if !cursor.goto_next_sibling() {
                errors.push(not_implemented_error(
                    "PropertyDeclaration::parse::goto_next_sibling",
                    node.parent().unwrap(),
                ));
                cursor.goto_parent();
                return Ok((None, errors));
            }

            let node = cursor.node();
            match node.kind() {
                "=" => (),
                "ERROR" => errors.push(parse_error(node)),
                _ => errors.push(not_implemented_error("PropertyDeclaration", node)),
            };
        }

        if !cursor.goto_next_sibling() {
            errors.push(not_implemented_error(
                "PropertyDeclaration::parse::goto_next_sibling",
                node.parent().unwrap(),
            ));
            cursor.goto_parent();
            return Ok((None, errors));
        }

        let (rhs, mut errs) = Expr::parse(cursor, source)?;
        errors.append(&mut errs);

        if cursor.goto_next_sibling() {
            let node = cursor.node();
            errors.push(not_implemented_error(
                "PropertyDeclaration::parse::goto_next_sibling",
                node,
            ));
        }

        cursor.goto_parent();

        match (qualifier, lhs, rhs) {
            (Some(qualifier), Some(lhs), Some(rhs)) => Ok((
                Some(Self {
                    qualifier,
                    lhs,
                    r#type,
                    rhs: Box::new(rhs),
                }),
                errors,
            )),
            _ => Ok((None, errors)),
        }
    }
}

fn parse_error(node: tree_sitter::Node) -> UserParseError {
    UserParseError {
        range: node.range(),
        message: "Invalid Swift".to_string(),
        context: node.kind().to_string(),
    }
}

fn not_implemented_error(context: &str, node: tree_sitter::Node) -> UserParseError {
    UserParseError {
        range: node.range(),
        message: "Not implemented".to_string(),
        context: format!("{context}::{kind}", kind = node.kind()),
    }
}
