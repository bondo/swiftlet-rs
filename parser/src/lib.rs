use pest::{
    iterators::Pair,
    pratt_parser::{Assoc, Op, PrattParser},
    Parser,
};

use ast::*;

#[derive(pest_derive::Parser)]
#[grammar = "grammar.pest"]
struct SwiftletParser;

pub fn parse(input: &str) -> Result<Program, SwiftletError> {
    let pairs = SwiftletParser::parse(Rule::Program, input).map_err(|e| {
        let span = match e.location {
            pest::error::InputLocation::Pos(pos) => pest::Span::new(input, pos, pos),
            pest::error::InputLocation::Span((start, end)) => pest::Span::new(input, start, end),
        };
        SwiftletError::syntax(span.unwrap(), e.variant.message())
    })?;

    let mut statements = vec![];

    for pair in pairs {
        match pair.as_rule() {
            Rule::Stmt => statements.extend(build_statements_from_pair(pair)),
            Rule::EOI => {}
            r => panic!("Unexpected rule in program: {r:?}"),
        }
    }

    Ok(Program { statements })
}

fn build_statements_from_pair(pair: Pair<Rule>) -> Vec<Statement> {
    debug_assert_eq!(pair.as_rule(), Rule::Stmt);
    pair.as_span().as_str().trim_end().to_string();

    let mut pairs = pair.into_inner();
    let pair = pairs.next().unwrap();

    let stmts = match pair.as_rule() {
        Rule::Assign => vec![Statement::Assignment(build_assignment_from_pair(pair))],
        Rule::PropertyDecl => build_property_decelerations_from_pair(pair)
            .into_iter()
            .map(Statement::PropertyDeceleration)
            .collect(),
        Rule::StructDecl => {
            let mut pairs = pair.into_inner();
            let name = build_identifier_from_pair(pairs.next().unwrap());
            let properties = pairs
                .flat_map(build_property_decelerations_from_pair)
                .collect();

            vec![Statement::StructDeceleration(StructDeceleration {
                name,
                properties,
            })]
        }
        Rule::If => {
            let mut pairs = pair.into_inner();
            let condition = build_spanned_expression_from_pair(pairs.next().unwrap());
            let if_branch = build_statements_from_pair(pairs.next().unwrap());
            let else_branch = if let Some(pair) = pairs.next() {
                build_statements_from_pair(pair)
            } else {
                vec![]
            };

            debug_assert_eq!(pairs.next(), None);

            vec![Statement::IfStatement(IfStatement {
                condition,
                if_branch,
                else_branch,
            })]
        }
        Rule::While => {
            let mut pairs = pair.into_inner();
            let condition = build_spanned_expression_from_pair(pairs.next().unwrap());
            let body = build_statements_from_pair(pairs.next().unwrap());

            debug_assert_eq!(pairs.next(), None);

            vec![Statement::WhileStatement(WhileStatement {
                condition,
                body,
            })]
        }
        Rule::Print => {
            let mut pairs = pair.into_inner();
            let expression = build_expression_from_pair(pairs.next().unwrap());

            debug_assert_eq!(pairs.next(), None);

            vec![Statement::PrintStatement(PrintStatement { expression })]
        }
        Rule::Block => {
            let statements = pair
                .into_inner()
                .flat_map(build_statements_from_pair)
                .collect();
            vec![Statement::BlockStatement(BlockStatement { statements })]
        }
        r => panic!("Unexpected rule in statement: {r:?}"),
    };

    debug_assert_eq!(pairs.next(), None);

    stmts
}

fn build_assignment_from_pair(pair: Pair<Rule>) -> Assignment {
    debug_assert_eq!(pair.as_rule(), Rule::Assign);

    let mut pairs = pair.into_inner();
    let lhs = build_identifier_from_pair(pairs.next().unwrap());
    let rhs = build_expression_from_pair(pairs.next().unwrap());

    debug_assert_eq!(pairs.next(), None);

    Assignment { lhs, rhs }
}

fn build_property_decelerations_from_pair(pair: Pair<Rule>) -> Vec<PropertyDeceleration> {
    debug_assert_eq!(pair.as_rule(), Rule::PropertyDecl);

    let mut pairs = pair.into_inner();
    let qualifier = match pairs.next().unwrap().as_rule() {
        Rule::Var => Qualifier::Var,
        Rule::Let => Qualifier::Let,
        v => panic!("Unexpected qualifier value: '{v:?}'"),
    };

    let mut decls = vec![];

    while let Some(pair) = pairs.next() {
        let name = build_identifier_from_pair(pair);

        let ty = {
            if let Some(Rule::Ident) = pairs.peek().map(|p| p.as_rule()) {
                Some(build_identifier_from_pair(pairs.next().unwrap()))
            } else {
                None
            }
        };

        let initializer = {
            if let Some(Rule::Expr) = pairs.peek().map(|p| p.as_rule()) {
                Some(build_expression_from_pair(pairs.next().unwrap()))
            } else {
                None
            }
        };

        decls.push(PropertyDeceleration {
            qualifier: qualifier.clone(),
            name,
            ty,
            initializer,
        });
    }

    decls
}

fn build_spanned_expression_from_pair(pair: Pair<Rule>) -> Spanned<Expression> {
    (pair.as_span(), build_expression_from_pair(pair))
}

fn build_expression_from_pair(pair: Pair<Rule>) -> Expression {
    debug_assert_eq!(pair.as_rule(), Rule::Expr);

    let mut pairs = pair.into_inner().rev();

    let mut expression = build_or_expression_from_pair(pairs.next().unwrap());

    while let Some(pair) = pairs.next() {
        let if_branch = build_or_expression_from_pair(pair);

        let condition = build_spanned_or_expression_from_pair(pairs.next().unwrap());

        expression = Expression::Conditional(
            Box::new(condition),
            Box::new(if_branch),
            Box::new(expression),
        )
    }

    expression
}

fn build_spanned_or_expression_from_pair(pair: Pair<Rule>) -> Spanned<Expression> {
    (pair.as_span(), build_or_expression_from_pair(pair))
}

fn build_or_expression_from_pair(pair: Pair<Rule>) -> Expression {
    debug_assert_eq!(pair.as_rule(), Rule::OrExpr);

    let mut pairs = pair.into_inner();

    let mut expr = build_and_expression_from_pair(pairs.next().unwrap());

    while let Some(op) = pairs.next() {
        debug_assert_eq!(op.as_rule(), Rule::OrOp);

        let rhs = build_and_expression_from_pair(pairs.next().unwrap());

        expr = Expression::Or(Box::new(expr), op.as_span(), Box::new(rhs));
    }

    expr
}

fn build_and_expression_from_pair(pair: Pair<Rule>) -> Expression {
    debug_assert_eq!(pair.as_rule(), Rule::AndExpr);

    let mut pairs = pair.into_inner();

    let mut expr = build_eq_expression_from_pair(pairs.next().unwrap());

    while let Some(op) = pairs.next() {
        debug_assert_eq!(op.as_rule(), Rule::AndOp);

        let rhs = build_eq_expression_from_pair(pairs.next().unwrap());

        expr = Expression::And(Box::new(expr), op.as_span(), Box::new(rhs));
    }

    expr
}

fn build_eq_expression_from_pair(pair: Pair<Rule>) -> Expression {
    debug_assert_eq!(pair.as_rule(), Rule::EqExpr);

    let mut pairs = pair.into_inner();

    let mut expr = build_cmp_expression_from_pair(pairs.next().unwrap());

    if let Some(pair) = pairs.next() {
        let op = match pair.as_rule() {
            Rule::Eq => EqualityOperator::Equal,
            Rule::Ne => EqualityOperator::NotEqual,
            r => panic!("Unexpected equality operator: {r:?}"),
        };

        let rhs = build_cmp_expression_from_pair(pairs.next().unwrap());

        expr = Expression::Equality(Box::new(expr), (pair.as_span(), op), Box::new(rhs));
    }

    expr
}

fn build_cmp_expression_from_pair(pair: Pair<Rule>) -> Expression {
    debug_assert_eq!(pair.as_rule(), Rule::CmpExpr);

    let mut pairs = pair.into_inner();

    let mut expr = build_pratt_expression_from_pair(pairs.next().unwrap());

    if let Some(pair) = pairs.next() {
        let op = match pair.as_rule() {
            Rule::Lt => ComparisonOperator::LessThan,
            Rule::Le => ComparisonOperator::LessThanOrEqual,
            Rule::Gt => ComparisonOperator::GreaterThan,
            Rule::Ge => ComparisonOperator::GreaterThanOrEqual,
            r => panic!("Unexpected comparison operator: {r:?}"),
        };

        let rhs = build_pratt_expression_from_pair(pairs.next().unwrap());

        expr = Expression::Comparison(Box::new(expr), (pair.as_span(), op), Box::new(rhs));
    }

    expr
}

fn build_pratt_expression_from_pair(pair: Pair<Rule>) -> Expression {
    debug_assert_eq!(pair.as_rule(), Rule::PrattExpr);

    let pratt = PrattParser::new()
        .op(Op::infix(Rule::Add, Assoc::Left) | Op::infix(Rule::Sub, Assoc::Left))
        .op(Op::infix(Rule::Mul, Assoc::Left)
            | Op::infix(Rule::Div, Assoc::Left)
            | Op::infix(Rule::Rem, Assoc::Left))
        .op(Op::prefix(Rule::Neg) | Op::prefix(Rule::Not));

    pratt
        .map_primary(|primary| match primary.as_rule() {
            Rule::Int => Expression::IntegerLiteral(primary.as_str().parse().unwrap()),
            Rule::Str => Expression::StringLiteral({
                let s = primary.as_str();
                s[1..s.len() - 1].to_string()
            }),
            Rule::True => Expression::BooleanLiteral(true),
            Rule::False => Expression::BooleanLiteral(false),
            Rule::Call => {
                let mut pairs = primary.into_inner();

                let expr_pair = pairs.next().unwrap();
                let mut expr_span = expr_pair.as_span();
                let mut expr = Expression::Identifier(build_identifier_from_pair(expr_pair));

                for pair in pairs {
                    let pair_span = pair.as_span();
                    match pair.as_rule() {
                        Rule::Args => {
                            let mut pairs = pair.into_inner();
                            let mut arguments = vec![];

                            while let Some(p) = pairs.peek() {
                                let label = if p.as_rule() == Rule::Ident {
                                    Some(build_identifier_from_pair(pairs.next().unwrap()))
                                } else {
                                    None
                                };

                                let value =
                                    build_spanned_expression_from_pair(pairs.next().unwrap());

                                arguments.push(Argument { label, value });
                            }

                            expr = Expression::Call(Box::new((expr_span, expr)), arguments);
                        }
                        Rule::Ident => {
                            let property = build_identifier_from_pair(pair);
                            expr = Expression::Navigation(Box::new((expr_span, expr)), property)
                        }
                        r => panic!("Unexpected rule in call expression: {:?}", r),
                    };

                    expr_span =
                        Span::new(expr_span.get_input(), expr_span.start(), pair_span.end())
                            .unwrap();
                }

                expr
            }
            Rule::Expr => build_expression_from_pair(primary),
            r => panic!("Unexpected rule in primary expression: {r:?}"),
        })
        .map_prefix(|op, rhs| match op.as_rule() {
            Rule::Neg => Expression::Unary((op.as_span(), UnaryOperator::Negate), Box::new(rhs)),
            Rule::Not => Expression::Unary((op.as_span(), UnaryOperator::Not), Box::new(rhs)),
            r => panic!("Unexpected rule in prefix expression: {r:?}"),
        })
        .map_infix(|lhs, op, rhs| match op.as_rule() {
            Rule::Add => Expression::Term(
                Box::new(lhs),
                (op.as_span(), TermOperator::Plus),
                Box::new(rhs),
            ),
            Rule::Sub => Expression::Term(
                Box::new(lhs),
                (op.as_span(), TermOperator::Minus),
                Box::new(rhs),
            ),
            Rule::Mul => Expression::Factor(
                Box::new(lhs),
                (op.as_span(), FactorOperator::Multiply),
                Box::new(rhs),
            ),
            Rule::Div => Expression::Factor(
                Box::new(lhs),
                (op.as_span(), FactorOperator::Divide),
                Box::new(rhs),
            ),
            Rule::Rem => Expression::Factor(
                Box::new(lhs),
                (op.as_span(), FactorOperator::Modulo),
                Box::new(rhs),
            ),
            r => panic!("Unexpected rule in infix expression: {r:?}"),
        })
        .parse(pair.into_inner())
}

fn build_identifier_from_pair(pair: Pair<Rule>) -> Spanned<String> {
    debug_assert_eq!(pair.as_rule(), Rule::Ident);
    (pair.as_span(), pair.as_str().to_string())
}

#[cfg(test)]
mod tests {
    use insta::assert_debug_snapshot;

    use super::*;

    #[test]
    fn expr_1() {
        let program =
            parse(r#"let foo = 5 || !"20" < true && (john - 4) * -2 ? 42 : 19 < 30 ? 14 : 9;"#)
                .unwrap_or_else(|e| panic!("\n{e}\n"));
        assert_debug_snapshot!(program);
    }

    #[test]
    fn expr_2() {
        let program = parse(r#"print(a > b && b < c);"#).unwrap_or_else(|e| panic!("\n{e}\n"));
        assert_debug_snapshot!(program);
    }

    #[test]
    fn example_1() {
        let program = parse(
            r#"
                var foo: Int = 4;
                let bar = foo;
                print(bar); // Prints 4
            "#,
        )
        .unwrap_or_else(|e| panic!("\n{e}\n"));
        assert_debug_snapshot!(program);
    }

    #[test]
    fn example_2() {
        let program = parse(
            r#"
                struct Vec2 { var x: Int, y: Int };
                var v = Vec2 (x: 4, y: 2);
                print(v.y); // Prints 2
            "#,
        )
        .unwrap_or_else(|e| panic!("\n{e}\n"));
        assert_debug_snapshot!(program);
    }
}
