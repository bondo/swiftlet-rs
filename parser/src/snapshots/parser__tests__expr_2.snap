---
source: parser/src/lib.rs
expression: program
---
Program {
    statements: [
        PrintStatement(
            PrintStatement {
                expression: And(
                    Comparison(
                        Identifier(
                            (
                                Span {
                                    str: "a",
                                    start: 6,
                                    end: 7,
                                },
                                "a",
                            ),
                        ),
                        (
                            Span {
                                str: ">",
                                start: 8,
                                end: 9,
                            },
                            GreaterThan,
                        ),
                        Identifier(
                            (
                                Span {
                                    str: "b",
                                    start: 10,
                                    end: 11,
                                },
                                "b",
                            ),
                        ),
                    ),
                    Span {
                        str: "&&",
                        start: 12,
                        end: 14,
                    },
                    Comparison(
                        Identifier(
                            (
                                Span {
                                    str: "b",
                                    start: 15,
                                    end: 16,
                                },
                                "b",
                            ),
                        ),
                        (
                            Span {
                                str: "<",
                                start: 17,
                                end: 18,
                            },
                            LessThan,
                        ),
                        Identifier(
                            (
                                Span {
                                    str: "c",
                                    start: 19,
                                    end: 20,
                                },
                                "c",
                            ),
                        ),
                    ),
                ),
            },
        ),
    ],
}
